### ISSUE TITLE
Panasonic DC-GH6 raws are version 8 - new decompressor needed

### ISSUE THREAD

**Issue by LebedevRI** [darktable-org/rawspeed#367]

**Comment by uvolfen** [comment]
I uploaded a full set (all variants of crops) on raw.pixls.us with Creative Commons Zero (CC0) license at August, 22nd.

Uwe

**Comment by Opendreas** [comment]
Any progress on adding GH6 to Darktable?

**Comment by yaiol** [comment]
Unfortunately last version of darktable 4.2.0 didn't add support to GH6, same with GH5M2 though files are available for both cameras on raw.pixls.us. 

Only way to use darktable with those cameras is to convert RAW pictures to DNG

**Comment by kmilos** [comment]
Tracker for other models depending on v8 decoder:

DC-S5M2: https://github.com/darktable-org/darktable/issues/13511
DC-S5M2X: https://github.com/darktable-org/darktable/issues/15446
DC-G9M2: https://github.com/darktable-org/darktable/issues/15679
DC-S9: https://github.com/darktable-org/darktable/issues/18609
DC-GH7
DC-S1RM2

**Comment by bringlein** [comment]
Any update on this?
Does it depend on `libraw` releasing the GH6 support as open source? (see [here](https://www.libraw.org/node/2710) and [here](https://www.libraw.org/comment/6626#comment-6626))? 
Thanks!

**Comment by kmilos** [comment]
> Does it depend on `libraw` releasing the GH6 support as open source?

It depends on _someone_ reverse engineering the new codec and releasing the spec or code as open source, it doesn't necessarily have to come from LibRaw.

**Comment by kmilos** [comment]
Just landed for porting: https://github.com/LibRaw/LibRaw/blob/master/src/decoders/pana8.cpp @jdneumeyer77 @mlouielu

**Comment by jdneumeyer77** [comment]
That code is pretty ugly. Not sure when I'll get a chance to work on it, but hopefully over the next couple weeks I'll take crack at an initial port.

**Comment by LebedevRI** [comment]
> That code is pretty ugly.

:)

> Not sure when I'll get a chance to work on it, but hopefully over the next couple weeks I'll take crack at an initial port.

If you do want to take a look at this, i would strongly suggest the PR should be *just* the code being copied over,
with no rewrites silencing compiler/clang-tidy diagnostics, or using rawspeed's native data structures.
That'll give us a known-"good" baseline change.

**Comment by LibRaw** [comment]
> That code is pretty ugly. Not sure when I'll get a chance to work on it, but hopefully over the next couple weeks I'll take crack at an initial port.

Unlike all other open source implementations of Panasonic RawFormat=8 decoders, this one works

**Comment by mlouielu** [comment]
> > That code is pretty ugly. Not sure when I'll get a chance to work on it, but hopefully over the next couple weeks I'll take crack at an initial port.
> 
> Unlike all other open source implementations of Panasonic RawFormat=8 decoders, this one works

Appreciate your team's work, and generously open sourced it in a swiftly manner.

Originally I thought it will be scheduled at the end of 2024.

**Comment by jdneumeyer77** [comment]
> Unlike all other open source implementations of Panasonic RawFormat=8 decoders, this one works

I never doubted it worked, since it's had time to incubate. I still appreciate the time and effort it took to reverse engineer and test it from your team. Thanks for releasing as open source.

**Comment by cytrinox** [comment]
I've finished my Rust port of the pana8 decoder: https://github.com/dnglab/dnglab/pull/437/commits/03c3d5dbeb654435bf461016c0a0297788b5cdee, based on libraw code.

**Comment by kenjohanson2** [comment]
Upvoting. I'd love to see Darktable support this V8 codec, **used by the GH6, but also I believe the G9M2** (which I have and can test). And I suspect the new GH7.. And I see G9M2 samples present at https://raw.pixls.us/data/Panasonic/DC-G9M2/. Thank you!

PS -is there some ETA/version of DT that will get this?

**Comment by LibRaw** [comment]
> see Darktable support this V8 codec

Darktable already uses LibRaw to process Canon CR3 files. So, adding such a processing path for Panasonic/V8 should be easy.
This will probably allow you to work with this file format much faster.

**Comment by kjyager** [comment]
**EDIT**: I've learned this is the wrong issue/tracker for this comment. I've truncated it so it isn't as much of an eyesore. 

New Darktable user and Lumix G9II owner here. I've taken a pass at adding support via libraw, and things are looking pretty good so far. I'm not very familiar with the codebase, so I've mostly just been muddling through. I'd love feedback from experienced contributors on what else ought to be done before submitting a PR, and for some guidelines regarding the amount of testing and consistency expected for contributions.

All I've done so far is add 'rw2' to the extension whitelist for libraw:
`imageio_libraw:274: const gchar *always_by_libraw = "cr3,rw2";`. 

and update the G9M2 entry in `cameras.xml`.

I may adopt some of the changes from [here](https://github.com/darktable-org/rawspeed/pull/529/files) as well.

Values for the black & white levels and color matrix are extracted from the EXIF on DNG's converted from my RW2 files, as documented by https://www.darktable.org/2012/10/whats-involved-with-adding-support-for-new-cameras/. 

This alone seems to have gotten me 95%+ of the way there. With these changes the RW2 files can be opened and edited. Opening both the RW2s and DNGs converted from them show them apparently matching with a just few caveats.

**EDIT**: [Original comment continued with info regarding testing of RW2's via this libraw workaround]

**Comment by kmilos** [comment]
> All I've done so far is add 'rw2' to the extension whitelist for libraw

Thanks, but that's not the real fix being looked for here.

P.S. The change to `cameras.xml` is irrelevant as everything is handled by LibRaw in that case. Furthermore, no code change (to **_darktable_** - we're in the _RawSpeed_ repo and tracker here) is necessary, as you can simply add "rw2" to the (undocumented) `libraw_extensions` line in your `darktablerc` config file to get to the same (unsupported) workaround.

**Comment by kjyager** [comment]
> > All I've done so far is add 'rw2' to the extension whitelist for libraw
> 
> Thanks, but that's not the real fix being looked for here.
> 
> ...

Oh, I see. Apologies for the confusion. I was scouring for info on RW2v8 support for DT, and ended up thinking this was the main issue tracking the topic. I see now this is the wrong repo. 

I'm still interested in pitching in if someone is willing to point me in the right direction. Seems like a lot of people are eagerly awaiting support for RW2v8. Is the stance that libraw's decoder not considered suitable for a release, thus requiring a full port/rewrite to rawspeed instead? Or does it make sense to push the libraw workaround further within the Darktable repo?

**Comment by kmilos** [comment]
> I'm still interested in pitching in if someone is willing to point me in the right direction. ... Is the stance that libraw's decoder not considered suitable for a release, thus requiring a full port/rewrite to rawspeed instead?

Great. See the [comment above](https://github.com/darktable-org/rawspeed/issues/367#issuecomment-2028316669).

> Or does it make sense to push the libraw workaround further within the Darktable repo?

IMHO it does not as it can readily be enabled from the user config file.

~~Edit: Sorry, I take it back - perhaps there is some tweaking to do the user config workaround (gives me a black image), but that's a separate topic for the other repo indeed.~~

Edit 2: Take that back again! :wink: It does simply work, only just up to G9M2, but not for the S9 that is also not yet supported by the public LibRaw 202403 snapshot...

**Comment by kjyager** [comment]
I've finally been able to make the time to work on this, and I've successfully implemented the V8 decompressor, having based it off the Rust implementation in @cytrinox's dnglab. 

It is working, but I have some refactoring to do, and I suspect I have not done everything needed to fully _add support_ for the G9II and the other version 8 cameras. Since I'm not a regular contributor and don't know what all additional changes are needed I figured I'd drop in here and ask what else needs to be done. 

Any pointers would be appreciated. 

Here is my current TODO list:
- [x] Create and implement `PanasonicV8Decompressor.{h,cpp}`
    - [x] Test on Panasonic G9II sample files
    - [x] Test on Panasonic S5II sample files
    - [ ] Test on Panasonic S1RII sample files
    - [ ] Test on Panasonic GH7 sample files
    - [ ] Test on Panasonic S9 sample files 
    - [ ] Others...? 
- [x] Add DC-G9M2 entries to cameras.xml including color-matrix from dng converter
- [ ] Add entries for other cameras using RW2v8 w/ color-matrices from dng converter
- [ ] Verify raw black and white points for sample files from all cameras 
- [ ] More...?

Additional goals for after basic support is merged:
- [ ] Metadata based lens distortion correction
- [ ] Enhanced color-matrix for G9II

**Comment by kmilos** [comment]
That's great news, thanks for working on this!

From your testing list only the GH6 is missing AFAICT. Note that I have [outstanding PRs for cameras.xml entries](https://github.com/darktable-org/rawspeed/pulls?q=is%3Apr+is%3Aopen+label%3Aincomplete) for all known models (S1RM2 is not final yet), so you can maybe skip those tasks and check those instead (I think those are preferred to be handled separately to the decoder code).

One important thing missing from the todos is verifying crops: either via manual `<Crop>` entries and/or implementing default vendor crop from metadata (and deciding whether to drop the manual ones then).

Lens correction and enhanced matrix are out of scope for RawSpeed, and those can be followed up on the dt issue tracker.

Other than that, I guess start a PR as soon as possible so @LebedevRI can start reviewing, with a hope there is chance to make it for [the June dt release](https://github.com/darktable-org/darktable/wiki/Release-cycle)?

**Comment by kjyager** [comment]
> I have [outstanding PRs for cameras.xml entries](https://github.com/darktable-org/rawspeed/pulls?q=is%3Apr+is%3Aopen+label%3Aincomplete) for all known models ..., so you can maybe skip those tasks and check those instead...

Thanks for pointing this out, I'll gladly defer to those. 

> PR as soon as possible so [@LebedevRI](https://github.com/LebedevRI) can start reviewing, with a hope there is chance to make it for [the June dt release](https://github.com/darktable-org/darktable/wiki/Release-cycle)?

Sounds good. I'll do my planned cleanup pass and then start a PR. June seems very achievable. 

Thanks for weighing in!

**Comment by ascertrobw** [comment]
Just an observation from trying with my DC-S9 today. The libraw_extensions=rw2 works to let me open the file. But the image is all black. Poking around, the White Balance being detected from the camera is way off (25,000). Can supply a test RW2 if useful.

 (don't see a separate tracker for it yet, happy to post there if I missed one).

**Comment by kjyager** [comment]
> Just an observation from trying with my DC-S9 today. The libraw_extensions=rw2 works to let me open the file. But the image is all black. Poking around, the White Balance being detected from the camera is way off (25,000). Can supply a test RW2 if useful.

Please do upload samples to https://raw.pixls.us/ if you could. 

I just tested my RW2 decompressor implementation on some samples I found elsewhere, and they also came out black unfortunately. Hopefully it is only a slight difference from the other recent Panasonic cameras and this can be fixed easily. I'll try to look into it, but it might have to wait until the change is released for the other models.

**Comment by kjyager** [comment]
> I just tested my RW2 decompressor implementation on some samples I found elsewhere, and they also came out black unfortunately. Hopefully it is only a slight difference from the other recent Panasonic cameras and this can be fixed easily. I'll try to look into it, but it might have to wait until the change is released for the other models.

Ok, at least partially a false alarm. Disabling all modules other than raw black/white point, demosaicing, and input/output color profiles revealed an image. The decompressor works, but clearly other tweaks need to be made to support the S9.

**Comment by ascertrobw** [comment]
> > Just an observation from trying with my DC-S9 today. The libraw_extensions=rw2 works to let me open the file. But the image is all black. Poking around, the White Balance being detected from the camera is way off (25,000). Can supply a test RW2 if useful.
> 
> Please do upload samples to https://raw.pixls.us/ if you could.
> 
> I just tested my RW2 decompressor implementation on some samples I found elsewhere, and they also came out black unfortunately. Hopefully it is only a slight difference from the other recent Panasonic cameras and this can be fixed easily. I'll try to look into it, but it might have to wait until the change is released for the other models.

Cool - will upload when site allows (currently says rebuilding). As an intermediate step I've added to DropBox if you wanted to grab them:

https://www.dropbox.com/scl/fo/1bueo8osh8ezn61bz4irk/AJDTmxHniZcOAU7eEfWBV-g?rlkey=7hcup7rvt9v62hocomohnlusc&dl=0

- P1000003 is a manual 28mm lens, so no lens info
- P1000007 is a Lumix 100mm macro, so it has lens info

Both images seem to behave the same, so the presence of lens info doesn't seem a factor.

**Comment by kmilos** [comment]
@ascertrobw LibRaw public snapshot supports only up to G9M2. Please discuss the LibRaw workaround elsewhere.

> Cool - will upload when site allows (currently says rebuilding).

That won't be for a while unfortunately. In the meantime, please [open a formal S9 support request](https://github.com/darktable-org/darktable/issues/new/choose), and share a link there to samples uploaded to some file sharing service.

**Comment by kmilos** [comment]
> As an intermediate step I've added to DropBox

Thank you for the samples! As mentioned, please still do open a separate request for this model so it can be tracked easily.

We'll also need samples for all the possible raw modes (high res etc.).

**Comment by ascertrobw** [comment]
> > As an intermediate step I've added to DropBox
> 
> Thank you for the samples! As mentioned, please still do open a separate request for this model so it can be tracked easily.
> 
> We'll also need samples for all the possible raw modes (high res etc.).

Done - I don't have any HHHR samples as yet. Will try and make some time to create 1 over the next few days - it's not a mode I ever really use myself.

**Comment by kjyager** [comment]
I've just put up the pull request for my decompressor. It is pull request #790. I imagine several rounds of feedback and changes will be needed before it gets approved.
